update-v1.c

```diff
@@ -0,0 +1,376 @@
+	memcpy(&ftr->revision, &data[6], 2);
+	ftr->flags = data[8];
+	ftr->misc = data[9];
+	ftr->footer_version = data[10];
+	memcpy(&ftr->magic, &data[11], 11);
+
+	if (strncmp("TS_UC_RA4M2", (char *)&ftr->magic, 11) != 0)
+		error(1, 1, "Invalid update file");
+
+	if (ftr->bin_size == 0 || ftr->bin_size > 128 * 1024)
+		error(1, 1, "Bin size is incorrect");
+
+	return 0;
+}
+int do_v1_micro_get_rev(board_t *board, int i2cfd, int *revision)
+{
+	*revision = speek16(i2cfd, board->i2c_chip, SUPER_REV_INFO);
+	*revision &= 0x7fff;
+
+	return 0;
+}
+
+int do_v1_micro_print_info(board_t *board, int i2cfd)
+{
+	uint16_t revision, modelnum;
+
+	modelnum = speek16(i2cfd, board->i2c_chip, SUPER_MODEL);
+	revision = speek16(i2cfd, board->i2c_chip, SUPER_REV_INFO);
+	revision &= 0x7fff;
+
+	printf("modelnum=0x%04X\n", modelnum);
+	printf("revision=%d\n", revision);
+	printf("dirty=%d\n", !!(revision & 0x8000));
+	return 0;
+}
+
+int do_v1_micro_get_file_rev(__attribute__((unused)) board_t *board, int *revision, char *update_path)
+{
+	struct micro_update_footer_v1 ftr;
+	int binfd;
+	int ret;
+
+	binfd = open(update_path, O_RDONLY | O_RSYNC);
+	if (binfd < 0)
+		error(1, errno, "Error opening update file");
+
+	ret = micro_update_parse_footer_v1(binfd, &ftr);
+	if (ret != 0)
+		return ret;
+
+	ret = close(binfd);
+	if (ret == -1)
+		return ret;
+
+	*revision = ftr.revision;
+
+	return 0;
+}
+
+int do_v1_micro_update(board_t *board, int i2cfd, char *update_path)
+{
+	uint16_t features0, status;
+	uint16_t crc;
+	uint32_t bin_size;
+	struct micro_update_footer_v1 ftr;
+	uint16_t buf[64];
+	int binfd;
+	int ret;
+	int i;
+	int retries = 10;
+
+	features0 = speek16(i2cfd, board->i2c_chip, SUPER_FEATURES0);
+	if ((features0 & SUPER_FEAT_FWUPD) == 0) {
+		fprintf(stderr, "The existing firmware does not support firmware updates. (0x%X)\n", features0);
+		return 1;
+	}
+
+	binfd = open(update_path, O_RDONLY | O_RSYNC);
+	if (binfd < 0)
+		error(1, errno, "Error opening update file");
+
+	micro_update_parse_footer_v1(binfd, &ftr);
+
+	if (ftr.model != board->modelnum) {
+		fprintf(stderr, "This update is for a %04X, not a %04X.\n", ftr.model, board->modelnum);
+		return 1;
+	}
+
+	/* gcc warns this pointer has alignment issues in packed structure. */
+	bin_size = (uint32_t)ftr.bin_size;
+
+	/* Check file is 128-byte aligned */
+	if (bin_size & 0x7F)
+		error(1, 0, "Binary file must be 128-byte aligned!");
+
+	fflush(stdout);
+
+	/*
+	 * Let the messages print out. Some of the flash operations will
+	 * cause the micro to drop some chars if they output while we touch 
+	 * flash 
+	 */
+	usleep(1000 * 10);
+
+	/* Write magic key and length/location information */
+	if (spokestream16(i2cfd, board->i2c_chip, SUPER_FL_MAGIC_KEY0, (uint16_t *)&magic_key, 4) < 0)
+		error(1, errno, "Failed to write magic key");
+
+	if (spokestream16(i2cfd, board->i2c_chip, SUPER_FL_SZ0, (uint16_t *)&bin_size, 4) < 0)
+		error(1, errno, "Failed to bin length");
+
+retry:
```

Comment:
Not something worth fixing since it would be a rewrite, but for the future this would be better to have as a function that would return on any of the `goto retry` points, and then this function would handle retry counts and calling that function in a loop.

Reply:
For now I removed these. I think they were needed with the very first firmware update code, but it has been a long time since those prototypes anyway. For new systems I'd prefer to know if we are still encountering situations where retry would be needed.